File: problems/CSES/Graph_Algorithms/distinct_routes/distinct_routes.py
import sys
from collections import deque
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

class Edge:
    def __init__(self, to, rev, cap, original=False):
        self.to = to
        self.rev = rev
        self.cap = cap
        self.original = original

def main():
    n, m = map(int, input().split())

    graph = [[] for _ in range(n)]

    def add_edge(u, v):
        forward = Edge(v, len(graph[v]), 1, True)
        backward = Edge(u, len(graph[u]), 0, False)

        graph[u].append(forward)
        graph[v].append(backward)

    for _ in range(m):
        a, b = map(int, input().split())
        add_edge(a - 1, b - 1)

    source = 0
    sink = n - 1

    level = [-1] * n
    ptr = [0] * n

    def bfs():
        level[:] = [-1] * n
        level[source] = 0

        q = deque([source])

        while q:
            u = q.popleft()

            for e in graph[u]:
                if e.cap > 0 and level[e.to] == -1:
                    level[e.to] = level[u] + 1
                    q.append(e.to)

        return level[sink] != -1

    def dfs(u, pushed):
        if u == sink:
            return pushed

        while ptr[u] < len(graph[u]):
            e = graph[u][ptr[u]]

            if e.cap > 0 and level[e.to] == level[u] + 1:
                flow = dfs(e.to, min(pushed, e.cap))

                if flow:
                    e.cap -= flow
                    graph[e.to][e.rev].cap += flow
                    return flow

            ptr[u] += 1

        return 0

    max_flow = 0

    while bfs():
        ptr[:] = [0] * n

        while True:
            pushed = dfs(source, 10**9)

            if not pushed:
                break

            max_flow += pushed

    def find_path():
        parent = [None] * n
        parent[source] = (-1, -1)

        q = deque([source])

        while q and parent[sink] is None:
            u = q.popleft()

            for i, e in enumerate(graph[u]):
                if (
                    e.original
                    and e.cap == 0
                    and parent[e.to] is None
                ):
                    parent[e.to] = (u, i)
                    q.append(e.to)

        path = []

        v = sink

        while v != source:
            u, edge_index = parent[v]

            path.append(v)

            e = graph[u][edge_index]
            e.cap = 1

            v = u

        path.append(source)
        path.reverse()

        return path

    print(max_flow)

    for _ in range(max_flow):
        path = find_path()

        print(len(path))
        print(*(v + 1 for v in path))

main()

File: problems/CSES/Graph_Algorithms/distinct_routes/distinct_routes.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

struct Edge {
    int to;
    int rev;
    int cap;
    bool original;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<vector<Edge>> graph(n);

    auto add_edge = [&](int u, int v) {
        Edge forward{v, (int)graph[v].size(), 1, true};
        Edge backward{u, (int)graph[u].size(), 0, false};

        graph[u].pb(forward);
        graph[v].pb(backward);
    };

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        a--;
        b--;

        add_edge(a, b);
    }

    int source = 0;
    int sink = n - 1;

    vector<int> level(n);
    vector<int> ptr(n);

    auto bfs = [&]() -> bool {
        fill(level.begin(), level.end(), -1);
        level[source] = 0;

        queue<int> q;
        q.push(source);

        while (!q.empty()) {
            int u = q.front();
            q.pop();

            for (const Edge &e : graph[u]) {
                if (e.cap > 0 && level[e.to] == -1) {
                    level[e.to] = level[u] + 1;
                    q.push(e.to);
                }
            }
        }

        return level[sink] != -1;
    };

    function<int(int, int)> dfs = [&](int u, int pushed) -> int {
        if (u == sink) return pushed;

        while (ptr[u] < (int)graph[u].size()) {
            Edge &e = graph[u][ptr[u]];

            if (e.cap > 0 && level[e.to] == level[u] + 1) {
                int flow = dfs(e.to, min(pushed, e.cap));

                if (flow) {
                    e.cap -= flow;
                    graph[e.to][e.rev].cap += flow;
                    return flow;
                }
            }

            ptr[u]++;
        }

        return 0;
    };

    int res = 0;
    while (bfs()) {
        fill(ptr.begin(), ptr.end(), 0);

        while (true) {
            int pushed = dfs(source, 1e9);

            if (!pushed) break;

            res += pushed;
        }
    }

    auto find_path = [&]() -> vector<int> {
        vector<pair<int, int>> parent(n, {-1, -1});
        vector<bool> visited(n, false);

        visited[source] = true;

        queue<int> q;
        q.push(source);

        while (!q.empty() && !visited[sink]) {
            int u = q.front();
            q.pop();

            for (int i = 0; i < (int)graph[u].size(); ++i) {
                Edge &e = graph[u][i];

                if (e.original && e.cap == 0 && !visited[e.to]) {
                    visited[e.to] = true;
                    parent[e.to] = {u, i};
                    q.push(e.to);
                }
            }
        }

        vector<int> path;

        int v = sink;
        while (v != source) {
            auto [u, edge_index] = parent[v];

            path.pb(v);

            Edge &e = graph[u][edge_index];
            e.cap = 1;

            v = u;
        }

        path.pb(source);
        reverse(path.begin(), path.end());

        return path;
    };

    cout << res << '\n';
    for (int i = 0; i < res; ++i) {
        vector<int> path = find_path();

        cout << path.size() << '\n';

        for (int v : path)
            cout << v + 1 << ' ';
        cout << '\n';
    }

    return 0;
}
